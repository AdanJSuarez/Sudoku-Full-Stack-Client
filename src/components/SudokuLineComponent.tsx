/** 
 * Adan J. Suarez
 * adanjsuarez@gmail.com
 * Full Stack Home-Project 
 */

import * as React from 'react';
import SudokuElementComponent from './SudokuElementComponent';

export interface ISudokuLineProps {
    row: number;
    dataLine: number[];
    toggledNumber: any;
}

export interface ISudokuLineState {
    currentRow  : number;
    currentToggleNumbers: any[];
    currentDataElement1: number | string;
    currentDataElement2: number | string;
    currentDataElement3: number | string;
    currentDataElement4: number | string;
    currentDataElement5: number | string;
    currentDataElement6: number | string;
    currentDataElement7: number | string;
    currentDataElement8: number | string;
    currentDataElement9: number | string;
}

export default class SudokuLineComponent extends React.Component<ISudokuLineProps, ISudokuLineState> {
    constructor(props: ISudokuLineProps) {
        super(props);
        this.state = {
            currentRow: props.row,
            currentToggleNumbers: [],
            currentDataElement1: props.dataLine[0],
            currentDataElement2: props.dataLine[1],
            currentDataElement3: props.dataLine[2],
            currentDataElement4: props.dataLine[3],
            currentDataElement5: props.dataLine[4],
            currentDataElement6: props.dataLine[5],
            currentDataElement7: props.dataLine[6],
            currentDataElement8: props.dataLine[7],
            currentDataElement9: props.dataLine[8],
        };
        this.toggleNumber = this.toggleNumber.bind(this);
    }

    public render() {
        return (
            <div className = 'Table-Line'>
                <table className='Table'>
                <tbody>
                    <tr>
                        <td className= 'td'> { <SudokuElementComponent row={this.props.row} column={1} dataElement={
                            this.state.currentDataElement1
                            } toggleNumber={this.toggleNumber}/> } </td>
                        <td className= 'td'> { <SudokuElementComponent row={this.props.row} column={2} dataElement={
                            this.state.currentDataElement2
                            } toggleNumber={this.toggleNumber}/> } </td>
                        <td className= 'td'> { <SudokuElementComponent row={this.props.row} column={3} dataElement={
                            this.state.currentDataElement3
                            } toggleNumber={this.toggleNumber}/> } </td>
                        <td></td>
                        <td className= 'td'> { <SudokuElementComponent row={this.props.row} column={4} dataElement={
                            this.state.currentDataElement4
                            } toggleNumber={this.toggleNumber}/> } </td>
                        <td className= 'td'> { <SudokuElementComponent row={this.props.row} column={5} dataElement={
                            this.state.currentDataElement5
                            } toggleNumber={this.toggleNumber}/> } </td>
                        <td className= 'td'> { <SudokuElementComponent row={this.props.row} column={6} dataElement={
                            this.state.currentDataElement6
                            } toggleNumber={this.toggleNumber}/> } </td>
                        <td></td>
                        <td className= 'td'> { <SudokuElementComponent row={this.props.row} column={7} dataElement={
                            this.state.currentDataElement7
                            } toggleNumber={this.toggleNumber}/> } </td>
                        <td className= 'td'> { <SudokuElementComponent row={this.props.row} column={8} dataElement={
                            this.state.currentDataElement8
                            } toggleNumber={this.toggleNumber}/> } </td>
                        <td className= 'td'> { <SudokuElementComponent row={this.props.row} column={9} dataElement={
                            this.state.currentDataElement9
                            } toggleNumber={this.toggleNumber}/> } </td>
                    </tr>
                </tbody>
                </table>
            </div>
        );
    }
    /**
     * Action before initial render. Hide numbers.
     *
     * @memberof SudokuLineComponent
     */
    public componentWillMount() {
        let hidenElements: number[] = this.selectHideElements();
            for (let i of hidenElements) {
                const element: string = "currentDataElement" + i.toString();
                this.state[element] = " ";
            }
    }
    /**
     *Take the Json object bubbled up
     * @param {*} element
     * @memberof SudokuLineComponent
     */
    public toggleNumber(element: any) {
      this.props.toggledNumber(element);
    }
    /**
     * Return a list of column to be hide.
     *
     * @private
     * @param {number} [max=4]
     * @returns {number[]}
     * @memberof SudokuLineComponent
     */
    private selectHideElements(max: number = 4): number[] {
        let result: number[] = [];
        for (let i = 0; i < max; i++) {
            let num: number = Math.floor(Math.random() * 9) + 1;
            result.push(num);
        }
        return result;
    }
}
