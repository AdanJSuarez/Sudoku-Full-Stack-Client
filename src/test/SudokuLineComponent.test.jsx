import React from 'react';
import Enzyme, {shallow, render, mount} from 'enzyme';
// import sinon from 'sinon';
import Adapter from 'enzyme-adapter-react-16';
import {createSerializer} from 'enzyme-to-json';
//import { expect } from 'chai';
import SudokuLineComponent from '../components/SudokuLine';

//Set the default serializer for jest to be from enzyme-to-json
//this produce a more easy to read serialized format
expect.addSnapshotSerializer(createSerializer({mode: "deep"}));
Enzyme.configure({adapter: new Adapter()});

describe("Test suit for SudokuLineComponent", ()=>{
    const dataLine = [2,5,8,7,4,3,9,6,1];
    
    it('render correctly shallow', ()=>{
        const wrapper = shallow(
            <SudokuLineComponent row={1} dataLine={dataLine} toggledNumber={{number:2, row:1, column:1}} />
        );
        expect(wrapper).toMatchSnapshot();
    });
    it('render correctly render', ()=>{
        const wrapper = render(
            <SudokuLineComponent row={1} dataLine={dataLine} toggledNumber={{number:2, row:1, column:1}} />
        );
        expect(wrapper).toMatchSnapshot();
    });
    it('render correctly mount', () => {
        const wrapper = mount(<SudokuLineComponent row={1} dataLine={dataLine} toggledNumber={{number:2, row:1, column:1}}/>);
        expect(wrapper).toMatchSnapshot();
    });
    
    //TODO: It needs many more test!!
})