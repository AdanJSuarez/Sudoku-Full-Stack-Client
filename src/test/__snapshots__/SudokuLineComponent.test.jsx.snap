// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suit for SudokuLineComponent render correctly mount 1`] = `
<div
  className="Table-Line"
>
  <table
    className="Table"
  >
    <tbody>
      <tr>
        <td
          className="td"
        >
           
          <button
            className="ElementSelected"
            onClick={[Function]}
          >
             
            2
             
          </button>
           
        </td>
        <td
          className="td"
        >
           
          <button
            className="ElementSelected"
            onClick={[Function]}
          >
             
            5
             
          </button>
           
        </td>
        <td
          className="td"
        >
           
          <button
            className="ElementSelected"
            onClick={[Function]}
          >
             
            8
             
          </button>
           
        </td>
        <td
          className="td"
        >
           
          <button
            className="ElementSelected"
            onClick={[Function]}
          >
             
            7
             
          </button>
           
        </td>
        <td
          className="td"
        >
           
          <button
            className="ElementSelected"
            onClick={[Function]}
          >
             
            4
             
          </button>
           
        </td>
        <td
          className="td"
        >
           
          <button
            className="ElementSelected"
            onClick={[Function]}
          >
             
            3
             
          </button>
           
        </td>
        <td
          className="td"
        >
           
          <button
            className="ElementSelected"
            onClick={[Function]}
          >
             
            9
             
          </button>
           
        </td>
        <td
          className="td"
        >
           
          <button
            className="ElementSelected"
            onClick={[Function]}
          >
             
            6
             
          </button>
           
        </td>
        <td
          className="td"
        >
           
          <button
            className="ElementSelected"
            onClick={[Function]}
          >
             
            1
             
          </button>
           
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Test suit for SudokuLineComponent render correctly render 1`] = `
<div
  class="Table-Line"
>
  <table
    class="Table"
  >
    <tbody>
      <tr>
        <td
          class="td"
        >
           
          <button
            class="ElementSelected"
          >
             2 
          </button>
           
        </td>
        <td
          class="td"
        >
           
          <button
            class="ElementSelected"
          >
             5 
          </button>
           
        </td>
        <td
          class="td"
        >
           
          <button
            class="ElementSelected"
          >
             8 
          </button>
           
        </td>
        <td
          class="td"
        >
           
          <button
            class="ElementSelected"
          >
             7 
          </button>
           
        </td>
        <td
          class="td"
        >
           
          <button
            class="ElementSelected"
          >
             4 
          </button>
           
        </td>
        <td
          class="td"
        >
           
          <button
            class="ElementSelected"
          >
             3 
          </button>
           
        </td>
        <td
          class="td"
        >
           
          <button
            class="ElementSelected"
          >
             9 
          </button>
           
        </td>
        <td
          class="td"
        >
           
          <button
            class="ElementSelected"
          >
             6 
          </button>
           
        </td>
        <td
          class="td"
        >
           
          <button
            class="ElementSelected"
          >
             1 
          </button>
           
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Test suit for SudokuLineComponent render correctly shallow 1`] = `
<div
  className="Table-Line"
>
  <table
    className="Table"
  >
    <tbody>
      <tr>
        <td
          className="td"
        >
           
          <SudokuElementComponent
            column={1}
            dataElement={2}
            row={1}
            toggleNumber={[Function]}
          />
           
        </td>
        <td
          className="td"
        >
           
          <SudokuElementComponent
            column={2}
            dataElement={5}
            row={1}
            toggleNumber={[Function]}
          />
           
        </td>
        <td
          className="td"
        >
           
          <SudokuElementComponent
            column={3}
            dataElement={8}
            row={1}
            toggleNumber={[Function]}
          />
           
        </td>
        <td
          className="td"
        >
           
          <SudokuElementComponent
            column={4}
            dataElement={7}
            row={1}
            toggleNumber={[Function]}
          />
           
        </td>
        <td
          className="td"
        >
           
          <SudokuElementComponent
            column={5}
            dataElement={4}
            row={1}
            toggleNumber={[Function]}
          />
           
        </td>
        <td
          className="td"
        >
           
          <SudokuElementComponent
            column={6}
            dataElement={3}
            row={1}
            toggleNumber={[Function]}
          />
           
        </td>
        <td
          className="td"
        >
           
          <SudokuElementComponent
            column={7}
            dataElement={9}
            row={1}
            toggleNumber={[Function]}
          />
           
        </td>
        <td
          className="td"
        >
           
          <SudokuElementComponent
            column={8}
            dataElement={6}
            row={1}
            toggleNumber={[Function]}
          />
           
        </td>
        <td
          className="td"
        >
           
          <SudokuElementComponent
            column={9}
            dataElement={1}
            row={1}
            toggleNumber={[Function]}
          />
           
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
